cmd/gen-key/rsa.go

```diff
@@ -0,0 +1,132 @@
+package main
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"log"
+	"math/big"
+
+	"github.com/miekg/pkcs11"
+)
+
+func rsaArgs(label string, modLen, exp uint) ([]*pkcs11.Mechanism, []*pkcs11.Attribute, []*pkcs11.Attribute) {
+	expSlice := make([]byte, 8)
+	binary.BigEndian.PutUint64(expSlice, uint64(exp))
+	log.Printf("\tEncoded public exponent (%d) as: %0X\n", exp, expSlice)
+	return []*pkcs11.Mechanism{
+			pkcs11.NewMechanism(pkcs11.CKM_RSA_PKCS_KEY_PAIR_GEN, nil),
+		},
+		[]*pkcs11.Attribute{
+			pkcs11.NewAttribute(pkcs11.CKA_LABEL, label),
+			pkcs11.NewAttribute(pkcs11.CKA_TOKEN, true),
+			pkcs11.NewAttribute(pkcs11.CKA_VERIFY, true),
+			pkcs11.NewAttribute(pkcs11.CKA_MODULUS_BITS, modLen),
+			pkcs11.NewAttribute(pkcs11.CKA_PUBLIC_EXPONENT, expSlice),
+		}, []*pkcs11.Attribute{
+			pkcs11.NewAttribute(pkcs11.CKA_LABEL, label),
+			pkcs11.NewAttribute(pkcs11.CKA_TOKEN, true),
+			// Prevent attributes being retrieved
+			pkcs11.NewAttribute(pkcs11.CKA_SENSITIVE, true),
+			// Prevent the key being extracted from the device
+			pkcs11.NewAttribute(pkcs11.CKA_EXTRACTABLE, false),
+			// Allow the key to sign data
+			pkcs11.NewAttribute(pkcs11.CKA_SIGN, true),
+		}
+}
+
+func rsaPub(ctx Ctx, session pkcs11.SessionHandle, object pkcs11.ObjectHandle, modLen, exp uint) (*rsa.PublicKey, error) {
+	// Retrieve the public exponent and modulus for the generated public key
+	attrs, err := ctx.GetAttributeValue(session, object, []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_PUBLIC_EXPONENT, nil),
+		pkcs11.NewAttribute(pkcs11.CKA_MODULUS, nil),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to retrieve key attributes: %s", err)
+	}
+
+	// Attempt to build the public key from the retrieved attributes
+	pubKey := &rsa.PublicKey{}
+	gotMod, gotExp := false, false
+	for _, a := range attrs {
+		switch a.Type {
+		case pkcs11.CKA_PUBLIC_EXPONENT:
+			pubKey.E = int(big.NewInt(0).SetBytes(a.Value).Int64())
+			// Check the provided public exponent was used
+			if pubKey.E != int(exp) {
+				return nil, errors.New("Returned CKA_PUBLIC_EXPONENT doesn't match expected exponent")
+			}
+			gotExp = true
+			log.Printf("\tPublic exponent: %d\n", pubKey.E)
+		case pkcs11.CKA_MODULUS:
+			pubKey.N = big.NewInt(0).SetBytes(a.Value)
+			// Check the right length modulus was generated on the device
```

Comment:
Is there any possibility of this running into trouble if the first bit (or byte?) is zero? Are we guaranteed to have a first bit of 1 by PKCS#11? If so we should document.

Reply:
This sent me down a rabbit hole trying to figure out how `big.Int` it actually calculates the `BitLen`, interestingly it returns `i*_W + bits.Len(bytes[i])` where `i = len(bytes)-1` and `_W = word size in bits` and `bits.Len(x)` returns the minimum number of bits required to represent `x`.

Since CKA_MODULUS should only ever be a positive big endian encoded slice (type CK_BYTE) according to the spec I think this means this should never be an issue...